### PR TITLE
[5.5] Blade Extension Class Support

### PR DESCRIPTION
**Pull request by paulredmond**
A simple blade extension system to laravel/framework which allows a Blade extension class to be defined and registered in the container. This will also allow you to inject services from the container easily.

I mentioned it in laravel/internals#682

Basically this allows you to group related Blade directives in a class and through tagging the Laravel framework could register all the blade directives and conditionals automatically. Registered providers could then tag an extension and Laravel would wire it up in blade.

Here's a basic example of what an extension class might look like:

```php
<?php

namespace App\Blade;

use Illuminate\Contracts\View\BladeExtension;

/**
 * Example blade extension class for grouping extensions
 */
class ExampleBladeExtension implements BladeExtension
{
    public function getDirectives()
    {
        return [
            'truncate' => [$this, 'truncate']
        ];
    }

    public function getConditionals()
    {
        return [
            'prod' => [$this, 'isProduction']
        ];
    }

    public function isProduction()
    {
        return app()->environment('production');
    }

    /**
     * @see https://zaengle.com/blog/exploring-laravels-custom-blade-directives
     */
    public function truncate($expression)
    {
        list($string, $length) = explode(',',str_replace(['(',')',' '], '', $expression));

        return "<?php echo e(strlen({$string}) > {$length} ? substr({$string},0,{$length}).'...' : {$string}); ?>";
    }
}
```

You can create it with `php artisan make:blade ExampleBladeExtension`.

Here's how you'd wire it up in your provider:

```php
// Service container available to extensions
$this->app->bind(ExampleExtension::class, function ($app) {
    return new ExampleExtension($app['foo']);
});

$this->app->tag([ExampleExtension::class], 'blade.extension');
```

### CODE REVIEW

File: src/Illuminate/Foundation/Console/BladeExtensionMakeCommand.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?php
+
+namespace Illuminate\Foundation\Console;
+
+use Illuminate\Support\Str;
+use Illuminate\Console\GeneratorCommand;
+use Symfony\Component\Console\Input\InputOption;
```
Comment (m1guelpf) [inline src/Illuminate/Foundation/Console/BladeExtensionMakeCommand.php]: You can remove this import

File: src/Illuminate/Foundation/Console/BladeExtensionMakeCommand.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+        return sprintf("%sExtension", $name);
+    }
+
+    /**
+     * Get the console command options.
+     *
+     * @return array
+     */
+    protected function getOptions()
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Console/BladeExtensionMakeCommand.php]: Useless?

File: src/Illuminate/Foundation/Providers/ArtisanServiceProvider.php [hunk 3]
```diff
@@ -125,6 +126,7 @@ class ArtisanServiceProvider extends ServiceProvider
      */
     protected $devCommands = [
         'AppName' => 'command.app.name',
+        'BladeExtensionMake' => 'command.blade.make',
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Providers/ArtisanServiceProvider.php]: This should be sorted and come after `AuthMake`

File: src/Illuminate/View/ViewServiceProvider.php [hunk 4]
```diff
@@ -133,4 +143,17 @@ public function registerBladeEngine($resolver)
             return new CompilerEngine($this->app['blade.compiler']);
         });
     }
+
+    public function defineCustomBladeExtensions()
```
Comment (lucasmichot) [inline src/Illuminate/View/ViewServiceProvider.php]: Doclocks?

File: src/Illuminate/Foundation/Providers/ArtisanServiceProvider.php [hunk 5]
```diff
@@ -204,6 +206,13 @@ protected function registerAuthMakeCommand()
         });
     }
 
+    protected function registerBladeExtensionMakeCommand()
```
Comment (lucasmichot) [inline src/Illuminate/Foundation/Providers/ArtisanServiceProvider.php]: Docblocks?

**Comment by taylorotwell** [comment]
Hmm, don't think I want to include this for now.

**Pull request closed without merging**